edge-hub/watchdog/linux/src/main.rs

```diff
@@ -0,0 +1,131 @@
+use std::sync::Arc;
+use std::thread;
+use std::time::Duration;
+use tracing::{error, info, subscriber, Level};
+use tracing_subscriber::fmt::Subscriber;
+
+fn main() -> Result<(), Error> {
+    block_on(async_main())
+}
+
+async fn async_main() -> Result<(), Error> {
+    init_logging();
+    info!("Starting watchdog");
+
+    let should_shutdown = match register_sigterm_listener() {
+        Ok(should_shutdown) => should_shutdown,
+        Err(e) => {
+            error!("Failed to register sigterm listener. Shutting down.");
+            return Err(e);
+        }
+    };
+
+    // start edgehub and blow up if can't start
+    let mut edgehub = Command::new("dotnet")
+        .arg("/app/Microsoft.Azure.Devices.Edge.Hub.Service.dll")
+        .stdout(Stdio::inherit())
+        .spawn()
+        .expect("Failed to start Edge Hub process");
+    info!("Launched Edge Hub process with pid {:?}", edgehub.id());
+
+    // unwrap broker if started, else shutdown edgehub and exit
+    let maybe_broker = match Command::new("/usr/local/bin/mqttd")
+        .stdout(Stdio::inherit())
+        .spawn()
+    {
+        Ok(child) => Some(child),
+        Err(e) => {
+            error!("Failed to start MQTT Broker process. {:?}", e);
+            None
+        }
+    };
+    let mut broker = maybe_broker.unwrap_or_else(|| {
+        info!("Broker process not started, so shutting down EdgeHub and exiting");
+        let shutdown = shutdown(&mut edgehub);
+        block_on(shutdown);
+        exit(1);
+    });
+    info!("Launched MQTT Broker process");
+
+    while !should_shutdown.load(Ordering::Relaxed) {
+        if !is_running(&mut edgehub) {
+            break;
+        }
+
+        if !is_running(&mut broker) {
+            break;
+        }
+
+        thread::sleep(Duration::from_secs(1));
+    }
+
+    let shutdown_edgehub = shutdown(&mut edgehub);
+    let shutdown_broker = shutdown(&mut broker);
+    join!(shutdown_edgehub, shutdown_broker);
+
+    info!("Stopped");
+    Ok(())
+}
+
+fn init_logging() {
+    let subscriber = Subscriber::builder().with_max_level(Level::INFO).finish();
+    let _ = subscriber::set_global_default(subscriber);
+}
+
+fn register_sigterm_listener() -> Result<Arc<AtomicBool>, Error> {
+    let signals = Signals::new(&[SIGTERM, SIGINT])?;
+    let should_shutdown = Arc::new(AtomicBool::new(false));
+    let sigterm_listener = should_shutdown.clone();
+    thread::spawn(move || {
+        for _sig in signals.forever() {
+            info!("Received shutdown signal for watchdog");
+            sigterm_listener.store(true, Ordering::Relaxed);
+        }
+    });
+
+    Ok(should_shutdown)
+}
+
+fn is_running(child_process: &mut Child) -> bool {
+    match child_process.try_wait() {
+        Ok(status) => status.is_none(),
+        Err(e) => {
+            error!("Error while polling child process. {:?}", e);
+            false
+        }
+    }
+}
+
+async fn shutdown(mut child: &mut Child) {
```

Comment:
I dont see any async code in this function. SHould we remove async than?

Reply:
Yes so I brought this up with Vadim. This runs synchronously, but we were considering replacing the call to thread::sleep() with some async thing that yielded back to the caller. Since we are removing async though I can remove.